pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentMetadataPushJobRunner.java

```diff
@@ -91,43 +131,123 @@ public void run() {
+    }
+
+    TableConfig tableConfig =
+        SegmentGenerationUtils.getTableConfig(_spec.getTableSpec().getTableConfigURI(), _spec.getAuthToken());
+    boolean consistentPushEnabled = ConsistentDataPushUtils.consistentDataPushEnabled(tableConfig);
+
+    Map<URI, String> uriToLineageEntryIdMap = null;
+    if (consistentPushEnabled) {
+      List<String> segmentNames = getSegmentsToReplace(segmentsToPushUriToTarPathMap);
+      uriToLineageEntryIdMap = ConsistentDataPushUtils.preUpload(_spec, segmentNames);
+    }
 
     int pushParallelism = _spec.getPushJobSpec().getPushParallelism();
     if (pushParallelism < 1) {
       pushParallelism = segmentsToPush.size();
     }
+
     if (pushParallelism == 1) {
       // Push from driver
-      try {
-        SegmentPushUtils.pushSegments(_spec, outputDirFS, segmentsToPush);
-      } catch (RetriableOperationException | AttemptsExceededException e) {
-        throw new RuntimeException(e);
+      if (consistentPushEnabled) {
+        SegmentPushUtils.sendSegmentUriAndMetadata(_spec, PinotFSFactory.create(outputDirURI.getScheme()),
+                segmentsToPushUriToTarPathMap);
+        try {
+          ConsistentDataPushUtils.postUpload(_spec, uriToLineageEntryIdMap);
+        } catch (Exception e) {
+          ConsistentDataPushUtils.handleUploadException(_spec, uriToLineageEntryIdMap, e);
+          throw new RuntimeException(e);
+        }
+      } else {
+        try {
+          SegmentPushUtils.pushSegments(_spec, outputDirFS, segmentsToPush);
+        } catch (RetriableOperationException | AttemptsExceededException e) {
+          throw new RuntimeException(e);
+        }
       }
     } else {
+      // Push from Spark executors when pushParallelism > 1
       JavaSparkContext sparkContext = JavaSparkContext.fromSparkContext(SparkContext.getOrCreate());
       JavaRDD<String> pathRDD = sparkContext.parallelize(segmentsToPush, pushParallelism);
       URI finalOutputDirURI = outputDirURI;
-      // Prevent using lambda expression in Spark to avoid potential serialization exceptions, use inner function
-      // instead.
-      pathRDD.foreach(new VoidFunction<String>() {
-        @Override
-        public void call(String segmentTarPath)
-            throws Exception {
+
+      if (consistentPushEnabled) {
+        // Parallel upload tasks executed by Spark executors
+        // In this case, we need to handle the case where the Spark job fails
+        sparkContext.sc().addSparkListener(new ConsistentDataPushFailureHandler(_spec, uriToLineageEntryIdMap));
+
+        pathRDD.foreach(segmentTarPath -> {
+          PluginManager.get().init();
+          for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
+            PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(),
+                new PinotConfiguration(pinotFSSpec));
+          }
+          Map<String, String> segmentUriToTarPathMap = SegmentPushUtils.getSegmentUriToTarPathMap(finalOutputDirURI,
```

Comment:
why are we recomputing the map? because it's running in spark context?

Reply:
this code is running inside the executor (inside the pathRDD split up into tasks) so it's computing a map relevant for the segments it got assigned to upload the metadata for